File: module-infra/src/main/java/maple/expectation/infrastructure/persistence/repository/RedisSessionRepositoryImpl.java
package maple.expectation.infrastructure.persistence.repository;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.Duration;
import java.time.Instant;
import java.util.HashSet;
import java.util.Set;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.domain.Session;
import maple.expectation.domain.repository.RedisSessionRepository;
import maple.expectation.infrastructure.executor.LogicExecutor;
import maple.expectation.infrastructure.executor.TaskContext;
import org.jspecify.annotations.Nullable;
import org.redisson.api.RMap;
import org.redisson.api.RedissonClient;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Repository;

/**
 * Redis 기반 세션 저장소 구현체
 *
 * <p>세션 구조 (Hash):
 *
 * <ul>
 *   <li>Key: session:{sessionId}
 *   <li>Fields: fingerprint, apiKey, myOcids (JSON), role, createdAt, lastAccessedAt
 *   <li>TTL: 30분 (Sliding Window)
 * </ul>
 */
@Slf4j
@Repository
public class RedisSessionRepositoryImpl implements RedisSessionRepository {

  private static final String SESSION_KEY_PREFIX = "session:";
  private static final String FIELD_FINGERPRINT = "fingerprint";
  private static final String FIELD_USER_IGN = "userIgn";
  private static final String FIELD_ACCOUNT_ID = "accountId";
  private static final String FIELD_API_KEY = "apiKey";
  private static final String FIELD_MY_OCIDS = "myOcids";
  private static final String FIELD_ROLE = "role";
  private static final String FIELD_CREATED_AT = "createdAt";
  private static final String FIELD_LAST_ACCESSED_AT = "lastAccessedAt";

  private final RedissonClient redissonClient;
  private final ObjectMapper objectMapper;
  private final LogicExecutor executor;
  private final long sessionTtlSeconds;

  public RedisSessionRepositoryImpl(
      RedissonClient redissonClient,
      ObjectMapper objectMapper,
      LogicExecutor executor,
      @Value("${auth.session.ttl}") long sessionTtlSeconds) {
    this.redissonClient = redissonClient;
    this.objectMapper = objectMapper;
    this.executor = executor;
    this.sessionTtlSeconds = sessionTtlSeconds;
  }

  /**
   * 세션을 저장합니다.
   *
   * @param session 저장할 세션
   */
  public void save(Session session) {
    String key = buildKey(session.sessionId());
    RMap<String, String> map = redissonClient.getMap(key);

    executor.executeVoidJava(
        () -> {
          map.put(FIELD_FINGERPRINT, session.fingerprint());
          map.put(FIELD_USER_IGN, session.userIgn());
          map.put(FIELD_ACCOUNT_ID, session.accountId());
          map.put(FIELD_API_KEY, session.apiKey());
          map.put(FIELD_MY_OCIDS, serializeOcids(session.myOcids()));
          map.put(FIELD_ROLE, session.role());
          map.put(FIELD_CREATED_AT, session.createdAt().toString());
          map.put(FIELD_LAST_ACCESSED_AT, session.lastAccessedAt().toString());

          // TTL 설정
          map.expire(Duration.ofSeconds(sessionTtlSeconds));
        },
        TaskContext.of("Session", "Save", session.sessionId()));

    log.debug("Session saved: sessionId={}, role={}", session.sessionId(), session.role());
  }

  /**
   * 세션 ID로 세션을 조회합니다.
   *
   * @param sessionId 세션 ID
   * @return 세션 (null if not found)
   */
  public @Nullable Session findById(String sessionId) {
    return executor.executeOrDefault(
        () -> doFindById(sessionId), null, TaskContext.of("Session", "FindById", sessionId));
  }

  private @Nullable Session doFindById(String sessionId) {
    String key = buildKey(sessionId);
    RMap<String, String> map = redissonClient.getMap(key);

    if (!map.isExists()) {
      return null;
    }

    String fingerprint = map.get(FIELD_FINGERPRINT);
    if (fingerprint == null) {
      return null;
    }

    return new Session(
        sessionId,
        fingerprint,
        map.get(FIELD_USER_IGN),
        map.get(FIELD_ACCOUNT_ID),
        map.get(FIELD_API_KEY),
        deserializeOcids(map.get(FIELD_MY_OCIDS)),
        map.get(FIELD_ROLE),
        Instant.parse(map.get(FIELD_CREATED_AT)),
        Instant.parse(map.get(FIELD_LAST_ACCESSED_AT)));
  }

  /**
   * 세션 TTL을 갱신합니다 (Sliding Window).
   *
   * @param sessionId 세션 ID
   * @return 갱신 성공 여부
   */
  public boolean refreshTtl(String sessionId) {
    return executor.executeOrDefault(
        () -> {
          String key = buildKey(sessionId);
          RMap<String, String> map = redissonClient.getMap(key);

          if (!map.isExists()) {
            return false;
          }

          map.put(FIELD_LAST_ACCESSED_AT, Instant.now().toString());
          map.expire(Duration.ofSeconds(sessionTtlSeconds));
          return true;
        },
        false,
        TaskContext.of("Session", "RefreshTtl", sessionId));
  }

  /**
   * 세션을 삭제합니다.
   *
   * @param sessionId 세션 ID
   */
  public void deleteById(String sessionId) {
    executor.executeVoidJava(
        () -> {
          String key = buildKey(sessionId);
          redissonClient.getMap(key).delete();
        },
        TaskContext.of("Session", "Delete", sessionId));

    log.debug("Session deleted: sessionId={}", sessionId);
  }

  /**
   * 세션 존재 여부를 확인합니다.
   *
   * @param sessionId 세션 ID
   * @return 존재 여부
   */
  public boolean existsById(String sessionId) {
    return executor.executeOrDefault(
        () -> {
          String key = buildKey(sessionId);
          return redissonClient.getMap(key).isExists();
        },
        false,
        TaskContext.of("Session", "Exists", sessionId));
  }

  private String buildKey(String sessionId) {
    return SESSION_KEY_PREFIX + sessionId;
  }

  /** OCIDs 직렬화 (CLAUDE.md Section 12 준수: LogicExecutor 패턴) */
  private String serializeOcids(Set<String> ocids) {
    if (ocids == null || ocids.isEmpty()) {
      return "[]";
    }
    return executor.executeOrDefault(
        () -> objectMapper.writeValueAsString(ocids),
        "[]",
        TaskContext.of("Session", "SerializeOcids", String.valueOf(ocids.size())));
  }

  /** OCIDs 역직렬화 (CLAUDE.md Section 12 준수: LogicExecutor 패턴) */
  private Set<String> deserializeOcids(String json) {
    if (json == null || json.isBlank()) {
      return new HashSet<>();
    }
    return executor.executeOrDefault(
        () -> objectMapper.readValue(json, new TypeReference<Set<String>>() {}),
        new HashSet<>(),
        TaskContext.of(
            "Session", "DeserializeOcids", json.length() > 20 ? json.substring(0, 20) : json));
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/persistence/repository/RedisSessionRepositoryImpl.kt
package maple.expectation.infrastructure.persistence.repository

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import maple.expectation.domain.Session
import maple.expectation.domain.repository.RedisSessionRepository as DomainRedisSessionRepository
import maple.expectation.infrastructure.executor.LogicExecutor
import maple.expectation.infrastructure.executor.TaskContext
import org.jspecify.annotations.Nullable
import org.redisson.api.RMap
import org.redisson.api.RedissonClient
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Repository
import java.time.Duration
import java.time.Instant

@Repository
open class RedisSessionRepositoryImpl(
    private val redissonClient: RedissonClient,
    private val objectMapper: ObjectMapper,
    private val executor: LogicExecutor,
    @Value("\${auth.session.ttl}") private val sessionTtlSeconds: Long,
) : DomainRedisSessionRepository {

    private companion object {
        const val SESSION_KEY_PREFIX = "session:"
        const val FIELD_FINGERPRINT = "fingerprint"
        const val FIELD_USER_IGN = "userIgn"
        const val FIELD_ACCOUNT_ID = "accountId"
        const val FIELD_API_KEY = "apiKey"
        const val FIELD_MY_OCIDS = "myOcids"
        const val FIELD_ROLE = "role"
        const val FIELD_CREATED_AT = "createdAt"
        const val FIELD_LAST_ACCESSED_AT = "lastAccessedAt"
    }

    override fun save(session: Session) {
        val key = buildKey(session.sessionId())
        val map = redissonClient.getMap<String, String>(key)

        executor.executeVoidJava(
            Runnable {
                map[FIELD_FINGERPRINT] = session.fingerprint()
                map[FIELD_USER_IGN] = session.userIgn()
                map[FIELD_ACCOUNT_ID] = session.accountId()
                map[FIELD_API_KEY] = session.apiKey()
                map[FIELD_MY_OCIDS] = serializeOcids(session.myOcids())
                map[FIELD_ROLE] = session.role()
                map[FIELD_CREATED_AT] = session.createdAt().toString()
                map[FIELD_LAST_ACCESSED_AT] = session.lastAccessedAt().toString()
                map.expire(Duration.ofSeconds(sessionTtlSeconds))
            },
            TaskContext.of("Session", "Save", session.sessionId()),
        )
    }

    override fun findById(sessionId: String): Session? {
        return executor.executeOrDefault(
            { doFindById(sessionId) },
            null,
            TaskContext.of("Session", "FindById", sessionId),
        )
    }

    private fun doFindById(sessionId: String): Session? {
        val key = buildKey(sessionId)
        val map = redissonClient.getMap<String, String>(key)

        if (!map.isExists) {
            return null
        }

        val fingerprint = map[FIELD_FINGERPRINT] ?: return null

        return Session(
            sessionId,
            fingerprint,
            map[FIELD_USER_IGN]!!,
            map[FIELD_ACCOUNT_ID]!!,
            map[FIELD_API_KEY]!!,
            deserializeOcids(map[FIELD_MY_OCIDS]),
            map[FIELD_ROLE]!!,
            Instant.parse(map[FIELD_CREATED_AT]!!),
            Instant.parse(map[FIELD_LAST_ACCESSED_AT]),
        )
    }

    override fun refreshTtl(sessionId: String): Boolean {
        return executor.executeOrDefault(
            {
                val key = buildKey(sessionId)
                val map = redissonClient.getMap<String, String>(key)

                if (!map.isExists) {
                    false
                } else {
                    map[FIELD_LAST_ACCESSED_AT] = Instant.now().toString()
                    map.expire(Duration.ofSeconds(sessionTtlSeconds))
                    true
                }
            },
            false,
            TaskContext.of("Session", "RefreshTtl", sessionId),
        )
    }

    override fun deleteById(sessionId: String) {
        executor.executeVoidJava(
            Runnable {
                val key = buildKey(sessionId)
                redissonClient.getMap<String, String>(key).delete()
            },
            TaskContext.of("Session", "Delete", sessionId),
        )
    }

    override fun existsById(sessionId: String): Boolean {
        return executor.executeOrDefault(
            {
                val key = buildKey(sessionId)
                redissonClient.getMap<String, String>(key).isExists
            },
            false,
            TaskContext.of("Session", "Exists", sessionId),
        )
    }

    private fun buildKey(sessionId: String): String = SESSION_KEY_PREFIX + sessionId

    private fun serializeOcids(ocids: Set<String>?): String {
        if (ocids.isNullOrEmpty()) {
            return "[]"
        }
        return executor.executeOrDefault(
            { objectMapper.writeValueAsString(ocids) },
            "[]",
            TaskContext.of("Session", "SerializeOcids", ocids.size.toString()),
        )
    }

    private fun deserializeOcids(json: String?): Set<String> {
        if (json.isNullOrBlank()) {
            return emptySet()
        }
        return executor.executeOrDefault(
            { objectMapper.readValue(json, object : TypeReference<Set<String>>() {}) },
            emptySet(),
            TaskContext.of("Session", "DeserializeOcids", json.take(20)),
        )
    }
}
